### DISCUSSION TITLE
How to connect to local rabbitmq server over AMQP deployed in Kubernetes

### DISCUSSION THREAD

**Discussion by ashfaqDell** [rabbitmq/rabbitmq-server#9540] [Other]
Hi there,

I am trying to connect to local rabbitmq server. I can easily access management UI using the URL http://localhost:15672/#/.  In order for above URL to work I exposed the rabbitmq pod using the below command.

`  kubectl -n default pod rabbitmq-server-0 --type LoadBalancer --port 15672`

But I can not access AMQP based rabbitmq service as it is not exposed. I don't know the way to expose it and make it public, so that I can access it using my program. 

When I try to connect using my C# program, it throws error.

```
ConnectionFactory factory = new ConnectionFactory();actory.UserName = "guest";
factory.Password = "guest";
factory.Uri = new Uri("AMQP://localhost:5672");

```
No connection could be made because the target machine actively refused it.

Is there a way to access both management UI on port 15672 and rabbitmq-service on 5672?

**Comment by michaelklishin** [comment]
This is a fairly generic Kubernetes question, there [isn't much specific to RabbitMQ](https://stackoverflow.com/questions/63334992/how-to-expose-kubernetes-hosted-rabbitmq-to-the-outside) if you use our Cluster Operator. If you DIY a deployment, [this blog explains what's involved](https://blog.rabbitmq.com/posts/2020/08/deploying-rabbitmq-to-kubernetes-whats-involved/) (we do not recommend DIY these days).

There are at least several options:

 * [Traefik](https://community.traefik.io/t/traefik-ingress-for-rabbitmq-amqp/19899), Nginx or a similar ingress controller extension
 * Good old services, one per port

"Target machine actively refused" is TCP speak for "port access was refused".

Also, you [cannot use user `guest` to access remote nodes](https://rabbitmq.com/access-control.html#default-state) for fairly obvious security reasons. You must [create a new user](https://rabbitmq.com/production-checklist.html#users-and-permissions) with credentials that are not known to the entire Internet.

**Comment by mkuratczyk** [comment]
If you skip the `--port` flag, all ports are exposed.

You can also define the service you want directly in the `RabbitmqCluster` definition https://www.rabbitmq.com/kubernetes/operator/using-operator.html#service-type
